# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1259609530 18000
#      Mon Nov 30 14:32:10 2009 -0500
# Node ID 1506a17832c95a7418a69305e49c9a114317eaf9
# Parent  483c4b09b788d92b1a0a0a43fc5c390e0bfd684e
doc building fixes for class methods

diff --git a/doc/ChangeLog b/doc/ChangeLog
--- a/doc/ChangeLog
+++ b/doc/ChangeLog
@@ -1,8 +1,17 @@
+2009-11-30  John W. Eaton  <jwe@octave.org>
+
+	* interpreter/munge-texi.cc (process_texi_input_file): Double @ in
+	symbol_name used in @anchor.
+	(extract_docstring): Double @ in function name used in @ref.
+
+	* interpreter/mk_doc_cache.m (cmd): Transform @ in class names to
+	@@ when processing DOCSTRINGS files.
+
 2009-11-30  John W. Eaton  <jwe@octave.org>
 
 	* interpreter/Makefile.a (nodist_octave_TEXINFOS): Add
 	contributors.texi and $(MUNGED_TEXI_SRC) to the list.
 	(dist_octave_TEXINFOS): Delete.
 	(DISTCLEANFILES, all_octave_TEXINFOS): New variables.
 	($(srcdir)/octave.info, octave.dvi, octave.pdf,	octave.html):
 	Depend on $(all_octave_TEXINFOS), not $(octave_TEXINFOS).
diff --git a/doc/interpreter/mk_doc_cache.m b/doc/interpreter/mk_doc_cache.m
--- a/doc/interpreter/mk_doc_cache.m
+++ b/doc/interpreter/mk_doc_cache.m
@@ -54,16 +54,17 @@ for i = 1:nfiles
     delim_idx = find (tmp == doc_delim, 1);
     text{i} = tmp(delim_idx:end);
   endif
 endfor
 text = [text{:}, doc_delim];
 
 text = regexprep (text, "@seealso *{([^}]*)}", "See also: $1.");
 text = regexprep (text, "-\\*- texinfo -\\*-[ \t]*[\r\n]*", "");
+text = regexprep (text, "@", "@@");
 
 [fid, name, msg] = mkstemp ("octave_doc_XXXXXX", true);
 
 if (fid < 0)
   error ("%s: %s\n", name, msg);
 endif
 
 fwrite (fid, text, "char");
diff --git a/doc/interpreter/munge-texi.cc b/doc/interpreter/munge-texi.cc
--- a/doc/interpreter/munge-texi.cc
+++ b/doc/interpreter/munge-texi.cc
@@ -122,16 +122,19 @@ extract_docstring (std::istream& is)
                           && c != '\n'
                           && c != ','
                           && c != '}');
                   if (first)
                     first = false;
                   else
                     doc += ", ";
 
+		  if (function_name[0] == '@')
+		    function_name = "@" + function_name;
+
                   doc += "@ref{doc-" + function_name + ",,"
 		    + function_name + "}";
 
                   // test for end of @seealso
                   if (c == '}') 
                     break;
                 }
               if (c == '}')
@@ -271,16 +274,19 @@ process_texi_input_file (std::istream& i
 
 			  while (isspace (doc_string[j]))
 			    j++;
 
 			  // Make `see also' references in functions
 			  // possible using @anchor{TAG} (new with
 			  // Texinfo 4.0).
 
+			  if (symbol_name[0] == '@')
+			    symbol_name = "@" + symbol_name;
+
 			  os << "@anchor{doc-" << symbol_name << "}\n";
 
 			  os << doc_string.substr (j);
 			}
 		      else
 			os << doc_string;
 		    }
 		}
diff --git a/doc/interpreter/system.txi b/doc/interpreter/system.txi
--- a/doc/interpreter/system.txi
+++ b/doc/interpreter/system.txi
@@ -247,25 +247,25 @@ about the status of files.
 @menu
 * FTP objects::
 * URL manipulation::
 @end menu
 
 @node FTP objects
 @subsection FTP objects
 
-@DOCSTRING(ftp)
+@DOCSTRING(@ftp/ftp)
 
-@DOCSTRING(mget)
+@DOCSTRING(@ftp/mget)
 
-@DOCSTRING(mput)
+@DOCSTRING(@ftp/mput)
 
-@DOCSTRING(ascii)
+@DOCSTRING(@ftp/ascii)
 
-@DOCSTRING(binary)
+@DOCSTRING(@ftp/binary)
 
 @node URL manipulation
 @subsection URL manipulation
 
 @DOCSTRING(urlread)
 
 @DOCSTRING(urlwrite)
 
diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,12 @@
+2009-11-30  John W. Eaton  <jwe@octave.org>
+
+	* mkdoc: Handle classes in function names.
+
 2009-11-27  David Bateman  <dbateman@free.fr>
 
 	* @ftp/ascii.m, @ftp/binary.m, @ftp/cd.m, @ftp/close.m,
 	@ftp/delete.m, @ftp/dir.m, @ftp/display.m, @ftp/ftp.m,
 	@ftp/loadobj.m, @ftp/mget.m, @ftp/mkdir.m, @ftp/module.mk,
 	@ftp/mput.m, @ftp/rename.m, @ftp/rmdir.m, @ftp/saveobj.m: New files
 	* Makefile.am: Add the @ftp to the build.
 
diff --git a/scripts/mkdoc b/scripts/mkdoc
--- a/scripts/mkdoc
+++ b/scripts/mkdoc
@@ -39,18 +39,22 @@ if test -f gethelp; then
 EOF
   for arg
   do
     if [ "$arg" = "--" ]; then
       prefix=""
     else
       echo "$prefix$arg" | $PERL -n -e 'chop;
 	$f = "$_";
-        next unless m{(.*)/(.*)\.m};
-	for (qx{./gethelp $2 "$f" < "$f"}) {
+        next unless m{(.*)/(@|)([^/]*)/(.*)\.m};
+        $fcn = "$4";
+        if ($2) {
+          $fcn = "$2$3/$fcn";
+        }
+	for (qx{./gethelp $fcn "$f" < "$f"}) {
 	  s/^\s+\@/\@/ unless $i_am_in_example;
 	  s/^\s+\@group/\@group/;
 	  s/^\s+\@end\s+group/\@end\s+group/;
 	  $i_am_in_example = 1 if /\s*\@example/;
 	  $i_am_in_example = 0 if /\s*\@end\s+example/;
 	  print;
 	}'
     fi
